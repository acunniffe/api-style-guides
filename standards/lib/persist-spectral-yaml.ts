import fsP from 'node:fs/promises';
import path from 'node:path';
import { RulesetDefinition } from '@stoplight/spectral-core';
import hash from 'node-object-hash';
import { ConfigAddressableSpectralRulesetDefinition } from './standard-definition';
const hasher = hash();
import { Collection, MongoClient } from 'mongodb';
import fs from 'fs';

export const localHashDir = path.resolve(process.cwd(), '.hashed-spectral');

export interface PersistSpectralYaml {
  persist(
    slug: string,
    content: RulesetDefinition,
    hostname?: string
  ): Promise<ConfigAddressableSpectralRulesetDefinition>;
  lookup(contentHash: string): Promise<RulesetDefinition | null>;
}

export class LocalPersist implements PersistSpectralYaml {
  constructor() {
    if (!fs.existsSync(localHashDir)) {
      fs.mkdirSync(localHashDir);
    }
  }
  async lookup(contentHash: string): Promise<RulesetDefinition | null> {
    try {
      const contents = (
        await fsP.readFile(path.join(localHashDir, contentHash))
      ).toString();
      return JSON.parse(contents) as unknown as RulesetDefinition;
    } catch (e) {
      return null;
    }
  }

  async persist(
    slug: string,
    content: RulesetDefinition,
    hostname: string = 'https://example.com'
  ): Promise<ConfigAddressableSpectralRulesetDefinition> {
    const name = hasher.hash(content).substring(0, 7);
    await fsP.writeFile(
      path.join(localHashDir, name),
      JSON.stringify(content, null, 2)
    );

    const url = new URL(hostname)
    url.pathname = `/api/${slug}/${name}`;

    return {
      url: url.toString(),
      contents: content,
      configHash: name,
    };
  }
}

export class CloudPersist implements PersistSpectralYaml {
  private configCollection: Collection<{ hash: string; value: any }>;

  constructor(password = process.env.MONGO_PASSWORD) {
    const uri = `mongodb+srv://backend:${password}@cluster0.yynrerb.mongodb.net/?retryWrites=true&w=majority`;
    const client = new MongoClient(uri);
    this.configCollection = client
      .db('configs')
      .collection<{ hash: string; value: any }>('config');
  }
  async lookup(contentHash: string): Promise<RulesetDefinition | null> {
    try {
      const lookup = await this.configCollection.findOne({ hash: contentHash });
      return lookup.value;
    } catch (e) {
      return null;
    }
  }

  async persist(
    slug: string,
    content: RulesetDefinition,
    hostname: string
  ): Promise<ConfigAddressableSpectralRulesetDefinition> {
    const hash = hasher.hash(content).substring(0, 7);

    try {
      await this.configCollection.insertOne({
        hash,
        value: content,
      });
    } catch (e) {
      if (!e.message.startsWith('E11000 duplicate key')) {
        throw new Error(e);
      } else {
        console.error(e);
      }
    }
    const url = new URL(hostname);
    url.pathname = `/api/${slug}/${hash}`;

    return {
      url: url.toString(),
      contents: content,
      configHash: hash,
    };
  }
}

//check prod
export const PersistedSpectralYaml: PersistSpectralYaml =
  process.env.NODE_ENV === 'production'
    ? new CloudPersist()
    : new LocalPersist();
