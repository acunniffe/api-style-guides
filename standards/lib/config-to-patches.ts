import { OpticYml, Patch, StandardDefinition } from './standard-definition';
import { RulesetDefinition } from '@stoplight/spectral-core';
import * as jsonPatch from 'fast-json-patch';
import { PersistedSpectralYaml } from './persist-spectral-yaml';

export const defaultOpticYml: OpticYml = Object.freeze({
  ruleset: [],
});

export const defaultSpectralYml: RulesetDefinition = Object.freeze({
  extends: [],
});

export type RunnerConfigs = {
  opticYml: OpticYml;
  spectralYml: RulesetDefinition;
  compatibility: {
    optic: boolean;
    spectral: boolean;
  };
};

export async function configToPatches(
  plugin: StandardDefinition<any>,
  globalConfig: Record<string, object>,
  runnerConfigs: RunnerConfigs = {
    compatibility: { optic: true, spectral: true },
    opticYml: jsonPatch.deepClone(defaultOpticYml),
    spectralYml: jsonPatch.deepClone(defaultSpectralYml),
  },
  hostname: string = 'https://example.com'
): Promise<RunnerConfigs> {
  const pluginActivated = globalConfig.hasOwnProperty(plugin.slug);
  if (pluginActivated) {
    const config = getConfig(plugin, globalConfig);

    const spectralGenerated = plugin.generateSpectralRuleset
      ? await plugin.generateSpectralRuleset(config)
      : undefined;

    let spectral;
    if (spectralGenerated) {
      spectral = await PersistedSpectralYaml.persist(
        plugin.slug,
        spectralGenerated,
        hostname
      );
    }

    const patches = plugin.toRulesetPatches(config, spectral);

    let updatedRunnerConfigs: RunnerConfigs =
      jsonPatch.deepClone(runnerConfigs);

    if (
      patches.compatibility.spectral &&
      canPatch(patches.compatibility.spectral.canPatch, config)
    ) {
      updatedRunnerConfigs.spectralYml = patchConfig(
        runnerConfigs.spectralYml,
        patches.compatibility.spectral.patch(runnerConfigs.spectralYml)
      );
    } else {
      updatedRunnerConfigs.compatibility.spectral = false;
    }

    if (
      patches.compatibility.optic &&
      canPatch(patches.compatibility.optic.canPatch, config)
    ) {
      updatedRunnerConfigs.opticYml = patchConfig(
        runnerConfigs.opticYml,
        patches.compatibility.optic.patch(runnerConfigs.opticYml)
      );
    } else {
      updatedRunnerConfigs.compatibility.optic = false;
    }

    return updatedRunnerConfigs;
  } else {
    return runnerConfigs;
  }
}

function getConfig(
  plugin: StandardDefinition<any>,
  globalConfig: Record<string, object>
) {
  const config = globalConfig[plugin.slug];
  const validate = plugin.isConfigValid(config);

  if (validate === true) {
    return config;
  } else {
    throw new Error(`${plugin.slug} config is not valid ${validate.join(',')}`);
  }
}

function canPatch(canPatch: boolean | ((config: any) => boolean), config: any) {
  if (typeof canPatch === 'boolean') return canPatch;
  return canPatch(config);
}

function patchConfig<YamlFormat extends OpticYml | RulesetDefinition>(
  configYaml: YamlFormat,
  patch: Patch<YamlFormat>
): YamlFormat {
  if ('errors' in patch) {
    throw new Error(patch.errors.join(', '));
  } else if ('updated' in patch) {
    return jsonPatch.deepClone(configYaml);
  } else if ('patch' in patch) {
    patch.patch;
    return jsonPatch.applyPatch(jsonPatch.deepClone(configYaml), patch.patch)
      .newDocument as unknown as YamlFormat;
  }
}
