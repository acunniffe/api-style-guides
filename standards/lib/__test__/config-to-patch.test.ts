import { it } from '@jest/globals';
import { configToPatches } from '../config-to-patches';

import { LocalPersist } from '../persist-spectral-yaml';
import BreakingChanges from '../../breaking-changes';
import UrlStyleGuide from '../../url-style-guides';

it('can create an optic.yml patch', async () => {
  const patches = await configToPatches(BreakingChanges, {
    [BreakingChanges.slug]: {
      exclude_operations_with_extension: 'x-draft',
    },
  });
  expect(patches).toMatchInlineSnapshot(`
    Object {
      "compatibility": Object {
        "optic": true,
        "spectral": false,
      },
      "opticYml": Object {
        "rulesets": Object {
          "breaking-changes": Object {
            "exclude_operations_with_extension": "x-draft",
          },
        },
      },
      "spectralYml": Object {
        "extends": Array [],
      },
    }
  `);
});
it('can create a spectral yml patch', async () => {
  const patches = await configToPatches(UrlStyleGuide, {
    [UrlStyleGuide.slug]: {
      'resource-names-plural': false,
    },
  });

  const lookup = await new LocalPersist().lookup(
    // @ts-ignore
    patches.spectralYml.extends[0]
  );
  expect(lookup).toMatchSnapshot();

  // this is wrong
  expect(patches).toMatchInlineSnapshot(`
    Object {
      "compatibility": Object {
        "optic": true,
        "spectral": true,
      },
      "opticYml": Object {
        "rulesets": Object {
          "spectral": Object {
            "added": Array [],
            "always": Array [
              "145a2f5",
            ],
          },
        },
      },
      "spectralYml": Object {
        "extends": Array [
          "145a2f5",
        ],
      },
    }
  `);
});
