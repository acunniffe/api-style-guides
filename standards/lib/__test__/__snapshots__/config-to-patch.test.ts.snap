// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can create a spectral yml patch 1`] = `
Object {
  "aliases": Object {
    "Request_Parameter_Path": Object {
      "description": "",
      "targets": Array [
        Object {
          "formats": Array [
            "oas3",
          ],
          "given": Array [
            "$.paths[*].parameters[?(@.in == 'path')].name",
            "$.paths[*][*].parameters[?(@.in == 'path')].name",
            "$.components.parameters[?(@.in == 'path')].name",
          ],
        },
        Object {
          "formats": Array [
            "oas2",
          ],
          "given": Array [
            "$.paths[*].parameters[?(@.in == 'path')].name",
            "$.paths[*][*].parameters[?(@.in == 'path')].name",
            "$.components.parameters[?(@.in == 'path')].name",
          ],
        },
      ],
    },
    "Request_Parameter_Query": Object {
      "description": "",
      "targets": Array [
        Object {
          "formats": Array [
            "oas3",
          ],
          "given": Array [
            "$.paths[*].parameters[?(@.in == 'query')].name",
            "$.paths[*][*].parameters[?(@.in == 'query')].name",
            "$.components.parameters[?(@.in == 'query')].name",
          ],
        },
        Object {
          "formats": Array [
            "oas2",
          ],
          "given": Array [
            "$.paths[*].parameters[?(@.in == 'query')].name",
            "$.paths[*][*].parameters[?(@.in == 'query')].name",
            "$.components.parameters[?(@.in == 'query')].name",
          ],
        },
      ],
    },
  },
  "formats": Array [
    "oas2",
    "oas3",
    "oas3.0",
    "oas3.1",
  ],
  "functionsDir": "functions",
  "rules": Object {
    "path-casing": Object {
      "description": "Paths must be \`kebab-case\`, with hyphens separating words.

**Invalid Example**

\`userInfo\` must be separated with a hyphen.

\`\`\`json
{
    \\"/userInfo\\": {
        \\"post: }
       ....
}
\`\`\` 

**Valid Example**

\`\`\`json
{
    \\"/user-info\\": {
       \\"post: }
       ....
}
\`\`\`",
      "given": Array [
        "$.paths",
      ],
      "message": "Paths must be kebab-case",
      "severity": "error",
      "then": Object {
        "field": "@key",
        "function": "pattern",
        "functionOptions": Object {
          "match": "^\\\\/([a-z0-9]+(-[a-z0-9]+)*)?(\\\\/[a-z0-9]+(-[a-z0-9]+)*|\\\\/{.+})*$",
        },
      },
    },
    "path-parameters-on-path-only": Object {
      "description": "Path parameters should be defined on the path level instead of the operation level.

**Invalid Example**

The \`user_id\` path parameter on line 8 should not be included with the \`patch' operation.

\`\`\`json lineNumbers
{      
  \\"patch\\": {
      \\"parameters\\": [
        {
          \\"schema\\": {
          \\"type\\": \\"integer\\"
         },
        \\"name\\": \\"user_id\\",
        \\"in\\": \\"path\\"
      }
    ]
  }
}
\`\`\`

**Valid Example**

The \`user-id\` path parameter is correctly located at the path level.

\`\`\`json lineNumbers

{  
    \\"paths\\": {
       \\"/users/{userId}\\": {
         \\"parameters\\": [
           {
             \\"schema\\": {
               \\"type\\": \\"integer\\"
           },
             \\"name\\": \\"user_id\\",
             \\"in\\": \\"path\\"
           }
       ]
     }
   }
 }
\`\`\`",
      "given": Array [
        "$.paths[*][*].parameters[?(@.in == 'path')]",
      ],
      "message": "Path parameters should be defined on the path level instead of the operation level.",
      "severity": "warn",
      "then": Object {
        "function": "falsy",
      },
    },
    "path-parameters-snake-case": Object {
      "description": "Path parameters must be \`snake_case\`, with each word separated by an underscore character and the first letter of each word lowercase. Also, the path parameter must not contain digits.

**Invalid Example**

The \`name\` property on line 9 (\`userId\`) must be separated by an underscore character and the \`I\` must be lowercase.

\`\`\`json lineNumbers
{
    \\"paths\\": {
      \\"/users/{userId}\\": {
        \\"parameters\\": [
          {
            \\"schema\\": {
              \\"type\\": \\"integer\\"
            },
            \\"name\\": \\"userId\\",
            \\"in\\": \\"path\\"
          }
        ]
      }
    }
  }
\`\`\`

**Valid Example**

\`\`\`json lineNumbers
{
    \\"paths\\": {
      \\"/users/{userId}\\": {
        \\"parameters\\": [
          {
            \\"schema\\": {
              \\"type\\": \\"integer\\"
            },
            \\"name\\": \\"user_id\\",
            \\"in\\": \\"path\\"
          }
        ]
      }
    }
  }
\`\`\`",
      "given": Array [
        "#Request_Parameter_Path",
      ],
      "message": "Path parameters must be snake_case and must not contain digits",
      "severity": "error",
      "then": Object {
        "field": "name",
        "function": "casing",
        "functionOptions": Object {
          "disallowDigits": true,
          "type": "snake",
        },
      },
    },
    "paths-avoid-special-characters": Object {
      "description": "Paths should not contain special characters, such as \`$\` \`&\` \`+\` \`,\` \`;\` \`=\` \`?\` and \`@%\`.

**Invalid Example**

The path contains an ampersand. 

\`\`\`json
{
    \\"/user&info\\": {
       \\"post: }
       ....
}
\`\`\` 

**Valid Example**

\`\`\`json
{
    \\"/user\\": {
       \\"post: }
       ....
}
\`\`\`",
      "given": Array [
        "$.paths",
      ],
      "message": "Avoid using special characters in paths",
      "severity": "warn",
      "then": Object {
        "field": "@key",
        "function": "pattern",
        "functionOptions": Object {
          "match": "",
          "notMatch": "^(.*)([\\\\$&+,;=?@%]+)(.*)$",
        },
      },
    },
    "paths-no-file-extensions": Object {
      "description": "Paths must not include \`json\` or \`xml\` file extensions.

**Invalid Example**

The path contains a \`.json\` extension. 

\`\`\`json
{
    \\"/user.json\\": {
       \\"post: }
       ....
}
\`\`\` 

**Valid Example**

\`\`\`json
{
    \\"/user\\": {
       \\"post: }
       ....
}
\`\`\`",
      "given": Array [
        "$.paths",
      ],
      "message": "Paths must not have file extensions",
      "severity": "error",
      "then": Object {
        "field": "@key",
        "function": "pattern",
        "functionOptions": Object {
          "notMatch": "\\\\b(JSON|json|XML|xml)\\\\b",
        },
      },
    },
    "paths-no-http-verbs": Object {
      "description": "Verbs such as \`get\`, \`delete\`, and \`put\` must not be included in paths because this information is conveyed by the HTTP method.

**Invalid Example**

The path contains the verb \`get\`. 

\`\`\`json
{
    \\"/getUsers\\": {
       \\"post: }
       ....
}
\`\`\` 

**Valid Example**

\`\`\`json
{
    \\"/user\\": {
       \\"post: }
       ....
}
\`\`\`",
      "given": Array [
        "$.paths",
      ],
      "message": "Paths must not have HTTP verbs in them",
      "severity": "error",
      "then": Object {
        "field": "@key",
        "function": "pattern",
        "functionOptions": Object {
          "notMatch": "\\\\b(GET|PUT|POST|DELETE|LIST|CREATE|REMOVE|get|put|post|delete|list|create|remove|Get|Put|Post|Delete|List|Create|Remove)\\\\b",
        },
      },
    },
    "paths-no-query-params": Object {
      "description": "Paths should not have query parameters in them. They should be defined separately in the OpenAPI.

**Invalid Example**

\`\`\`json
{
 \\"/users/{?id}\\": {

\`\`\`

**Valid Example**

\`\`\`json lineNumbers
{
    \\"parameters\\": [
        {
            \\"schema\\": {
                \\"type\\": \\"string\\"
            },
            \\"name\\": \\"id\\",
            \\"in\\": \\"path\\",
            \\"required\\": true,
            \\"description\\": \\"User's ID\\"
        }
    ]
}

\`\`\`",
      "given": Array [
        "$.paths",
      ],
      "message": "Paths should not have query parameters in them. They should be defined separately in the OpenAPI.",
      "severity": "warn",
      "then": Object {
        "field": "@key",
        "function": "pattern",
        "functionOptions": Object {
          "notMatch": "\\\\?",
        },
      },
    },
    "paths-no-trailing-slash": Object {
      "description": "Paths must not end with a trailing slash. 

\`/users\` and \`/users/\` are separate paths. It's considered bad practice for them to differ based only on a trailing slash. It's usually preferred to not have a trailing slash.

**Invalid Example**

The \`users\` path ends with a slash. 

\`\`\`json
{
    \\"/users/\\": {
       \\"post: }
       ....
}
\`\`\` 

**Valid Example**

\`\`\`json
{
    \\"/user\\": {
       \\"post: }
       ....
}
\`\`\`",
      "given": Array [
        "$.paths",
      ],
      "message": "Paths must not end with a trailing slash",
      "severity": "error",
      "then": Object {
        "field": "@key",
        "function": "pattern",
        "functionOptions": Object {
          "notMatch": "\\\\/$",
        },
      },
    },
    "query-parameters-snake-case": Object {
      "description": "Query parameters must be \`snake_case\`, with each word separated by an underscore character and the first letter of each word lowercase. Also, the query parameter must not contain digits.

**Invalid Example**

The \`name\` property on line 8 (\`user-Id\`) must be separated by an underscore character and the \`I\` must be lowercase.

\`\`\`json lineNumbers
{
   \\"parameters\\": [
     {
       \\"schema\\": {
         \\"type\\": \\"string\\"
       },
       \\"in\\": \\"query\\",
       \\"name\\": \\"user-Id\\"
     }
   ]
}
\`\`\`

**Valid Example**

\`\`\`json lineNumbers
{
    \\"parameters\\": [
      {
        \\"schema\\": {
          \\"type\\": \\"string\\"
        },
        \\"in\\": \\"query\\",
        \\"name\\": \\"user_id\\"
      }
    ]
 }
\`\`\`",
      "given": Array [
        "#Request_Parameter_Query",
      ],
      "message": "Query parameters should be snake_case and not contain digits",
      "severity": "error",
      "then": Object {
        "function": "casing",
        "functionOptions": Object {
          "disallowDigits": true,
          "type": "snake",
        },
      },
    },
    "server-has-api": Object {
      "description": "Server must have /api",
      "formats": Array [
        "oas3",
      ],
      "given": Array [
        "$.servers[*].url",
      ],
      "message": "Server must have /api in it",
      "severity": "info",
      "then": Object {
        "function": "pattern",
        "functionOptions": Object {
          "match": "^.*\\\\/api/?.*",
        },
      },
    },
    "server-lowercase": Object {
      "description": "Server URLs must be lowercase. This standard helps meet industry best practices.

**Invalid Example**

The \`name\` property on line 8 (\`user-Id\`) must be separated by an underscore character and the \`I\` must be lowercase.

\`\`\`json
{
    \\"servers\\": [
      {
        \\"url\\": \\"https://ACME.com/api\\"
      }
    ]
}
\`\`\`

**Valid Example**

\`\`\`json
{
    \\"servers\\": [
      {
        \\"url\\": \\"https://acme.com/api\\"
      }
    ]
}
\`\`\`",
      "formats": Array [
        "oas3",
      ],
      "given": Array [
        "$.servers[*].url",
      ],
      "message": "Server URL must be lowercase",
      "severity": "error",
      "then": Object {
        "function": "pattern",
        "functionOptions": Object {
          "match": "^[^A-Z]*$",
        },
      },
    },
  },
}
`;
